haskell/cabal.bzl

```diff
@@ -226,6 +226,33 @@ def _prepare_cabal_inputs(
             ],
             uniquify = True,
         )
+
+    # When building in a static context, we need to make sure that Cabal passes
+    # a couple of options that ensure any static code it builds can be linked
+    # correctly.
+    #
+    # * If we are using a static runtime, then we need to pass GHC both
+    #   `-fPIC` and `-fexternal-dynamic-refs`. With `-fPIC` alone, GHC still
+    #   generate `R_X86_64_PC32` relocations, which prevents loading its static
+    #   libraries as PIC.
+    #
+    # * If are building fully-statically-linked binaries, we need to ensure that
+    #   we pass arguments to `hsc2hs` such that objects it builds are
+    #   statically linked, otherwise we'll get dynamic linking errors when
+    #   trying to execute those objects to generate code as part of the build.
+    #   Since the static configuration should ensure that all the objects
+    #   involved are themselves statically built, this is just a case of passing
+    #   `-static` to the linker used by `hsc2hs` (which will be our own wrapper
+    #   script which eventually calls `gcc`, etc.).
+    if hs.toolchain.static_runtime:
+        args.add_all([
+            "--ghc-option=-fPIC",
+            "--ghc-option=-fexternal-dynamic-refs",
```

Comment:
See my other comment -- should we check `is_windows` here? I don't think the same applies for `fully_static_link` since at present using that with `is_windows` is just simply not supported, right?

Reply:
Indeed, `-fexternal-dynamic-refs` should be Unix only.